# Task 1: Data extraction and initial analysis

In [1]:
import pandas as pd

# Load data
file_path = 'financial_data.csv'
df = pd.read_csv(file_path)

# Print actual column names loaded by pandas to double-check
print("DataFrame columns after loading:")
print(df.columns)
print("-" * 30)

# Data Cleaning and Type Conversion 
# List of columns that should be numeric
numeric_cols = [
    'Total Revenue ($M)',
    'Net Income ($M)',
    'Total Assets ($M)',
    'Total Liabilities ($M)',
    'Cash Flow from Operating Activities ($M)'
]

# Apply cleaning and conversion to each numeric column
for col in numeric_cols:
    if col in df.columns: # Check if the column exists
        # Convert to string, remove commas, remove any leading/trailing spaces, then convert to numeric
        df[col] = pd.to_numeric(df[col].astype(str).str.replace(',', '').str.strip(), errors='coerce')
    else:
        print(f"Warning: Column '{col}' not found in DataFrame. Please check your CSV header names.")

# Check data types after conversion 
print("\nData Types after conversion:")
print(df[numeric_cols].dtypes)
print("-" * 30)

# Sort data before calculating percentage change 
# Ensure 'Fiscal Year' is sorted correctly for pct_change to work across years within each company
df = df.sort_values(by=['Company', 'Fiscal Year'])

# Calculate year-over-year changes 
df['Revenue Growth (%)'] = df.groupby('Company')['Total Revenue ($M)'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby('Company')['Net Income ($M)'].pct_change() * 100
df['Total Assets Growth (%)'] = df.groupby('Company')['Total Assets ($M)'].pct_change() * 100
df['Total Liabilities Growth (%)'] = df.groupby('Company')['Total Liabilities ($M)'].pct_change() * 100
df['Cash Flow from Operating Activities Growth (%)'] = df.groupby('Company')['Cash Flow from Operating Activities ($M)'].pct_change() * 100

# Fill NA values that result from pct_change calculations with 0 or an appropriate value
df.fillna(0, inplace=True)

# Display the DataFrame with new growth columns
print("\nDataFrame with Growth Metrics:")
df

DataFrame columns after loading:
Index(['Company', 'Fiscal Year', 'Total Revenue ($M)', 'Net Income ($M)',
       'Total Assets ($M)', 'Total Liabilities ($M)',
       'Cash Flow from Operating Activities ($M)'],
      dtype='object')
------------------------------

Data Types after conversion:
Total Revenue ($M)                          int64
Net Income ($M)                             int64
Total Assets ($M)                           int64
Total Liabilities ($M)                      int64
Cash Flow from Operating Activities ($M)    int64
dtype: object
------------------------------

DataFrame with Growth Metrics:


Company  Fiscal Year  Total Revenue ($M)  Net Income ($M)  \
8      Apple         2022              394328            99803   
7      Apple         2023              383285            96995   
6      Apple         2024              391035            93736   
2  Microsoft         2022              198270            72738   
1  Microsoft         2023              211915            72361   
0  Microsoft         2024              236584            78416   
5      Tesla         2022               81462            12587   
4      Tesla         2023               96773            14974   
3      Tesla         2024               97690             7153   

   Total Assets ($M)  Total Liabilities ($M)  \
8             352755                  302083   
7             352583                  290437   
6             364980                  308030   
2             364840                  198298   
1             411972                  205729   
0             484792                  253303   
5              82338                   36440   
4             106618                   43009   
3             122070                   48390   

   Cash Flow from Operating Activities ($M)  Revenue Growth (%)  \
8                                    122151            0.000000   
7                                    110543           -2.800461   
6                                    118254            2.021994   
2                                     89035            0.000000   
1                                     87582            6.882030   
0                                    109219           11.640988   
5                                     14724            0.000000   
4                                     13256           18.795267   
3                                     14923            0.947578   

   Net Income Growth (%)  Total Assets Growth (%)  \
8               0.000000                 0.000000   
7              -2.813543                -0.048759   
6              -3.359967                 3.516052   
2               0.000000                 0.000000   
1              -0.518299                12.918540   
0               8.367767                17.675959   
5               0.000000                 0.000000   
4              18.964010                29.488207   
3             -52.230533                14.492862   

   Total Liabilities Growth (%)  \
8                      0.000000   
7                     -3.855232   
6                      6.057424   
2                      0.000000   
1                      3.747390   
0                     23.124596   
5                      0.000000   
4                     18.026894   
3                     12.511335   

   Cash Flow from Operating Activities Growth (%)  
8                                        0.000000  
7                                       -9.502992  
6                                        6.975566  
2                                        0.000000  
1                                       -1.631942  
0                                       24.704848  
5                                        0.000000  
4                                       -9.970117  
3                                       12.575438

In [2]:
# Summarise these findings for each company

# Define the list of growth columns
growth_cols = ['Revenue Growth (%)', 'Net Income Growth (%)',
               'Total Assets Growth (%)', 'Total Liabilities Growth (%)',
               'Cash Flow from Operating Activities Growth (%)']

# Calculate and display Average Growth Rates per Company for all relevant metrics
print("\nAverage Growth Rates per Company (FY2022-2024):\n")
summary = df.groupby('Company')[growth_cols].mean()
summary


Average Growth Rates per Company (FY2022-2024):



Revenue Growth (%)  Net Income Growth (%)  Total Assets Growth (%)  \
Company                                                                         
Apple               -0.259489              -2.057837                 1.155764   
Microsoft            6.174339               2.616490                10.198166   
Tesla                6.580948             -11.088841                14.660357   

           Total Liabilities Growth (%)  \
Company                                   
Apple                          0.734064   
Microsoft                      8.957329   
Tesla                         10.179409   

           Cash Flow from Operating Activities Growth (%)  
Company                                                    
Apple                                           -0.842475  
Microsoft                                        7.690969  
Tesla                                            0.868440

# Financial Data Analysis for GFC AI Chatbot Project

## Data Loading and Preprocessing

Financial data was manually extracted from the latest available 10-K annual reports of Microsoft, Tesla, and Apple, directly from the SEC EDGAR database. This ensures high reliability and accuracy as these are audited financial statements. The following five core financial metrics were collected for Fiscal Years 2022, 2023, and 2024: Total Revenue, Net Income, Total Assets, Total Liabilities, and Cash Flow from Operating Activities.

The extracted data was organised into a CSV file and loaded into a pandas DataFrame. Crucially, columns containing numerical figures (such as 'Total Revenue ($M)') were cleaned by removing non-numeric characters (like commas) and then successfully converted to integer data types (`int64`), which was essential for accurate mathematical calculations and trend analysis.

---

## Key Findings by Company

### Apple
* **Performance:** A mature company showing relatively stable revenue but experiencing average slight declines in Net Income and Operating Cash Flow over the period. Revenue dipped in FY2023 but slightly recovered in FY2024. Net Income, however, saw consistent declines.
* **Growth Profile:** More modest average growth rates (some negative), reflecting its established market position. Assets and Liabilities showed minimal average growth.

### Microsoft
* **Performance:** Demonstrated strong and consistent financial performance. It achieved robust average revenue growth (6.17%) and healthy average growth across all other key metrics, including Net Income (2.62%) and Operating Cash Flow (7.69%).
* **Growth Profile:** Exhibited the most stable and balanced growth among the three companies, indicating strong operational execution and market demand.

### Tesla
* **Performance:** Characterised by rapid expansion in assets (average 14.66% growth) and liabilities (10.18%), coupled with strong average revenue growth (6.58%). However, it experienced significant volatility in Net Income, showing a large average decline (-11.09%), and variable Operating Cash Flow.
* **Growth Profile:** High-growth, but with notable swings in profitability, reflecting its dynamic and potentially more volatile industry landscape.

---

## Comparative Insights

* **Scale vs. Growth:** Apple and Microsoft operate on a significantly larger financial scale. Tesla, from a smaller base, shows higher asset and liability growth percentages indicative of aggressive expansion, but with more volatile profitability.
* **Profitability Trends:** Microsoft's profitability trends generally align with its revenue growth, while Apple saw declining net income despite stable revenue. Tesla's net income was highly volatile.
* **Cash Flow Generation:** All companies generate positive operating cash flow. Apple leads in absolute cash generated, while Microsoft shows strong, consistent growth in cash flow. Tesla's cash flow growth was more variable.

---

## Conclusion 

This analysis provides crucial insights for the GFC AI Chatbot. The chatbot can be designed to:
* **Identify company growth stages:** Distinguishing between mature (Apple), consistently growing (Microsoft), and high-growth/volatile (Tesla) profiles.
* **Analyse profitability and cash generation stability:** Flagging inconsistencies between revenue growth and net income/cash flow.
* **Facilitate comparative assessments:** Allowing users to quickly benchmark companies based on various financial metrics and growth trends, offering a nuanced view beyond just top-line numbers.

This foundational understanding of financial trends equips the chatbot with the intelligence to provide more insightful and actionable financial analysis.

# Task 2: Developing an AI-powered financial chatbot

In [3]:
import re # Import the regular expression module to help find years in text

In [4]:
def get_financial_insight(query, financial_df):
    """
    Processes a user query about financial data using rule-based logic and adds
    interactive suggestions for follow-up questions.
    """
    query = query.lower() # Convert query to lowercase for case-insensitive matching

    # --- 1. Identify Company ---
    # The assistant first tries to figure out which company you're asking about.
    company = None # Starts by assuming no company is identified yet.
    if "microsoft" in query:
        company = "Microsoft"
    elif "tesla" in query:
        company = "Tesla"
    elif "apple" in query:
        company = "Apple"

    if not company:
        return "I need a company name (Microsoft, Tesla, or Apple) to provide financial insights. Please try again."

    # Now that it knows the company, it goes to your big data table ('financial_df')
    # and pulls out *only* the rows that belong to that specific company.
    company_data = financial_df[financial_df['Company'] == company]

    # --- 2. Identify Specific Year in Query ---
    requested_year = None
    # This line uses a regular expression (re) to look for a four-digit number
    # that matches the years 2022, 2023, or 2024 within your question.
    # '\b' ensures it matches whole words (e.g., '2022' not just part of 'something2022').
    year_match = re.search(r'\b(202[2-4])\b', query) # Pattern specifically for 2022, 2023, 2024
    if year_match:
        # If a year is found, it converts it to a number.
        requested_year = int(year_match.group(1))

    # --- 3. Select Data for the Requested Year or Latest Year ---
    selected_year_data = None
    actual_year_used = None # This variable will store the year that the chatbot actually uses for its response.

    # Initialise response_parts here to ensure it always exists
    response_parts = []
    
    # If a specific year was asked for AND that year exists for the company in the data:
    if requested_year and requested_year in company_data['Fiscal Year'].values:
        # It picks the row that matches the exact requested year.
        selected_year_data = company_data[company_data['Fiscal Year'] == requested_year].iloc[0]
        actual_year_used = requested_year # The year used is the requested year.
    else:
        # If no year was asked for, or the requested year isn't in the data for that company,
        # it defaults to using the data from the latest available year (.iloc[-1]).
        selected_year_data = company_data.iloc[-1]
        actual_year_used = selected_year_data['Fiscal Year'] # The year used is the latest year.
        
        # If a year was requested but not found, it adds a polite message to the response.
        if requested_year:
            response_parts = [f"I couldn't find data for {company} in FY{requested_year}. Displaying data for FY{actual_year_used} instead. "]
        else:
            # If no specific year was requested, start with an empty list for response parts.
            response_parts = []

    # Use selected_year_data instead of latest_year_data for all metric lookups from now on.
    response_parts_temp = [] # A temporary list to build the main part of the response.
    metric_found = False # A flag to track if any specific metric was found in the query.
    
    # --- 4. Identify Metric and Construct Response ---
    # This section checks what specific financial number or growth rate you're asking for.

    # Check for absolute financial numbers (like total revenue, net income, etc.)
    if "total revenue" in query or "revenue" in query:
        value = selected_year_data['Total Revenue ($M)']
        response_parts_temp.append(f"{company}'s Total Revenue for FY{actual_year_used} was ${value:,.0f}M.")
        metric_found = True
    if "net income" in query or "profit" in query:
        value = selected_year_data['Net Income ($M)']
        response_parts_temp.append(f"{company}'s Net Income for FY{actual_year_used} was ${value:,.0f}M.")
        metric_found = True
    if "total assets" in query or "assets" in query:
        value = selected_year_data['Total Assets ($M)']
        response_parts_temp.append(f"{company}'s Total Assets for FY{actual_year_used} were ${value:,.0f}M.")
        metric_found = True
    if "total liabilities" in query or "liabilities" in query:
        value = selected_year_data['Total Liabilities ($M)']
        response_parts_temp.append(f"{company}'s Total Liabilities for FY{actual_year_used} were ${value:,.0f}M.")
        metric_found = True
    if "cash flow from operating activities" in query or "operating cash flow" in query or "cash flow" in query:
        value = selected_year_data['Cash Flow from Operating Activities ($M)']
        response_parts_temp.append(f"{company}'s Cash Flow from Operating Activities for FY{actual_year_used} was ${value:,.0f}M.")
        metric_found = True

    # Check for growth rates.
    # Growth percentages are always relative to the previous year available.
    # In our data, 2022 won't have growth, as there's no 2021 in the dataset.
    if len(company_data) > 1: # Only try to get growth if there's enough data for it.
        if "revenue growth" in query:
            value = selected_year_data['Revenue Growth (%)']
            # If the growth value is missing (NaN), it means there's no prior year to calculate from.
            if pd.isna(value):
                response_parts_temp.append(f"{company}'s Revenue Growth data for FY{actual_year_used} is not available (requires previous year's data).")
            else:
                response_parts_temp.append(f"{company}'s Revenue Growth for FY{actual_year_used} was {value:.2f}%.")
            metric_found = True
        if "net income growth" in query or "profit growth" in query:
            value = selected_year_data['Net Income Growth (%)']
            if pd.isna(value):
                 response_parts_temp.append(f"{company}'s Net Income Growth data for FY{actual_year_used} is not available (requires previous year's data).")
            else:
                response_parts_temp.append(f"{company}'s Net Income Growth for FY{actual_year_used} was {value:.2f}%.")
            metric_found = True
        if "assets growth" in query:
            value = selected_year_data['Total Assets Growth (%)']
            if pd.isna(value):
                response_parts_temp.append(f"{company}'s Total Assets Growth data for FY{actual_year_used} is not available (requires previous year's data).")
            else:
                response_parts_temp.append(f"{company}'s Total Assets Growth for FY{actual_year_used} was {value:.2f}%.")
            metric_found = True
        if "liabilities growth" in query:
            value = selected_year_data['Total Liabilities Growth (%)']
            if pd.isna(value):
                response_parts_temp.append(f"{company}'s Total Liabilities Growth data for FY{actual_year_used} is not available (requires previous year's data).")
            else:
                response_parts_temp.append(f"{company}'s Total Liabilities Growth for FY{actual_year_used} was {value:.2f}%.")
            metric_found = True
        if "operating cash flow growth" in query or "cash flow growth" in query:
            value = selected_year_data['Cash Flow from Operating Activities Growth (%)']
            if pd.isna(value):
                response_parts_temp.append(f"{company}'s Cash Flow from Operating Activities Growth data for FY{actual_year_used} is not available (requires previous year's data).")
            else:
                response_parts_temp.append(f"{company}'s Cash Flow from Operating Activities Growth for FY{actual_year_used} was {value:.2f}%.")
            metric_found = True

    # --- 5. Handle General Summaries or Unrecognised Queries & Add Interactivity ---

    # If you asked for a summary, performance, or overview.
    if "summarise" in query or "performance" in query or "overview" in query or "financial health" in query:
        # It creates a detailed summary response with several key financial points for the actual year used.
        summary_response = f"Here's a summary of {company}'s financial performance for FY{actual_year_used}:\n"
        summary_response += f"- Total Revenue: ${selected_year_data['Total Revenue ($M)']:,}M\n"
        summary_response += f"- Net Income: ${selected_year_data['Net Income ($M)']:,}M\n"
        summary_response += f"- Cash Flow from Operations: ${selected_year_data['Cash Flow from Operating Activities ($M)']:,}M\n"
        
        # It adds growth rates to the summary if they are available for the specific year.
        # Growth is only available from 2023 onwards in your dataset, as 2022 is the first year.
        if actual_year_used in [2023, 2024] and len(company_data) > 1:
            summary_response += f"- Revenue Growth (YoY): {selected_year_data.get('Revenue Growth (%)', 'N/A'):.2f}%\n"
            summary_response += f"- Net Income Growth (YoY): {selected_year_data.get('Net Income Growth (%)', 'N/A'):.2f}%\n"
        else:
             summary_response += f"Growth data for FY{actual_year_used} is not available (requires previous year's data in the dataset).\n"

        # After giving a summary, it suggests diving deeper or comparing years.
        summary_response += f"\nIs there a specific metric you'd like to dive deeper into, or perhaps compare another year's performance?"
        return summary_response

    # Combine any initial messages (like "year not found") with the metric-specific responses.
    response_parts.extend(response_parts_temp)

    # If specific metrics were found in the query.
    if response_parts:
        final_response = " ".join(response_parts)
        # It adds a follow-up question related to what you just asked.
        if "revenue" in query and "growth" not in query:
            final_response += f"\nWould you also like to know about {company}'s net income or revenue growth for FY{actual_year_used}?"
        elif "net income" in query and "growth" not in query:
            final_response += f"\nPerhaps {company}'s cash flow or net income growth for FY{actual_year_used} next?"
        elif "assets" in query or "liabilities" in query:
            final_response += f"\nWould you like a summary of {company}'s overall financial health for FY{actual_year_used}?"
        elif "growth" in query: # If any growth metric was asked
             final_response += f"\nWould you like to know about {company}'s other growth metrics or a summary of its financial health for FY{actual_year_used}?"
        return final_response
    else:
        # If the assistant couldn't find any specific metric or a summary request,
        # it gives you a polite message asking you to try rephrasing,
        # and reminds you what it can answer, including specific years.
        return (f"I'm not sure how to answer that about {company}. "
                f"I can tell you about its total revenue, net income, assets, liabilities, cash flow, or their growth rates. "
                f"Try asking 'What is Microsoft's revenue for 2023?' or 'Summarise Apple's performance for 2022'.")

# --- 6. Chatting with the Chatbot ---
# This is the part that runs the actual conversation in your Jupyter Notebook.

print("\n--- Financial Insights Chatbot ---")
# Engaging welcome message and examples.
print("Hello! I'm here to help you understand financial data for Microsoft, Tesla, and Apple.")
print("You can ask about their Total Revenue, Net Income, Total Assets, Total Liabilities, Cash Flow from Operating Activities, or their growth rates.")
print("You can also ask for specific years between 2022 and 2024. For example, 'What is Apple's revenue for 2022?'")
print("Try 'What is Microsoft's revenue growth?' or 'Summarise Apple's performance for 2023'.")
print("Type 'exit' to end our chat.")

# This is a loop that means: "Keep doing this forever, until I tell you to stop."
while True:
    user_input = input("\nYour query: ")

    # It checks if what you typed (all lowercase) is exactly "exit".
    if user_input.lower() == 'exit':
        print("Goodbye! It was a pleasure assisting you.")
        break # If you typed 'exit', this tells the loop to stop, ending the conversation.

    # This is the most important step: it takes your question ('user_input') and the
    # financial data table ('df') and sends them to the 'get_financial_insight' function (the chatbot's brain).
    # The answer that the function comes up with is saved in 'response'.
    response = get_financial_insight(user_input, df)

    # Finally, it prints the chatbot's answer back to you on the screen.
    print(f"Chatbot: {response}")


--- Financial Insights Chatbot ---
Hello! I'm here to help you understand financial data for Microsoft, Tesla, and Apple.
You can ask about their Total Revenue, Net Income, Total Assets, Total Liabilities, Cash Flow from Operating Activities, or their growth rates.
You can also ask for specific years between 2022 and 2024. For example, 'What is Apple's revenue for 2022?'
Try 'What is Microsoft's revenue growth?' or 'Summarise Apple's performance for 2023'.
Type 'exit' to end our chat.

Your query: Hi
Chatbot: I need a company name (Microsoft, Tesla, or Apple) to provide financial insights. Please try again.

Your query: What is the total revenue of Microsoft?
Chatbot: Microsoft's Total Revenue for FY2024 was $236,584M.
Would you also like to know about Microsoft's net income or revenue growth for FY2024?

Your query: What is the net income of Microsoft in 2022?
Chatbot: Microsoft's Net Income for FY2022 was $72,738M.
Perhaps Microsoft's cash flow or net income growth for FY2022 next?


## Chatbot Functionality Documentation

This section provides a summary of the AI-powered financial chatbot developed as part of Task 2 of the BCG Generative AI Virtual Internship. This chatbot leverages the financial data extracted and analysed in Task 1 to provide interactive insights.

### 1. Overview and Principles

The chatbot operates on a **rule-based logic** foundation. Its primary goal is to simulate a conversation with human users, understanding and responding to queries about financial data in a helpful and clear manner. My specific focus for this task was on implementing these core rule-based responses, while other team members handle more advanced NLP, ML, and UX aspects.

### 2. How the Chatbot Works

* **Keyword Detection:** The chatbot processes user queries by identifying specific keywords and patterns. It converts all input to lowercase to ensure case-insensitivity (e.g., "Microsoft" vs. "microsoft"), enhancing its ability to match user intent.
* **Company Identification:** It first determines the target company by looking for hardcoded names: Microsoft, Tesla, and Apple. This initial step is crucial for narrowing down the scope of the financial data to be retrieved.
* **Metric & Growth Identification:** Once a company is identified, the chatbot then scans the query for keywords related to specific financial metrics (e.g., "revenue," "net income," "assets," "liabilities," "cash flow from operating activities") or their corresponding year-over-year growth rates (e.g., "revenue growth," "net income growth"). It also utilises `re.search` function to precisely identify specific fiscal years (2022-2024) if mentioned.
* **Data Integration:** The chatbot directly queries the `pandas` DataFrame (`df`), which was prepared in Task 1. This DataFrame contains the cleaned absolute financial figures and pre-calculated year-over-year growth percentages for Fiscal Years 2022-2024. The data is retrieved by filtering for the identified company and, if applicable, the specified fiscal year.
* **Response Generation:** After retrieving the relevant data points, the chatbot constructs a clear, human-readable, and summarised response. It can provide specific metric values, growth percentages, or a general financial performance overview for a given company and year. The design also includes logic to suggest follow-up questions, enhancing user interaction and guiding further exploration of the financial data.

### 3. Data Structuring and Retrieval
#### Data Structuring:
Financial data is organised in a `pandas` DataFrame by company and fiscal year, including key metrics and pre-calculated Year-over-Year growth percentages. This structure ensures efficient access to both latest and historical records.

#### Retrieval Method:
Queries filter the DataFrame by identified company and, if specified, fiscal year. If a requested year is unavailable, it defaults to the most recent data. Retrieved numerical values are formatted for clarity.

### 4. Communicating Financial Insights
#### Dialogue Design:
The chatbot provides direct answers and proactively suggests relevant follow-up questions based on query context. Error handling includes graceful fallback messages when specific years are not found.

#### Example Interaction with the Chatbot:
Example 1: Absolute Metric for a Different Company and Year
- **User query:** What is the total revenue of Microsoft?
- **Chatbot response:** Microsoft's Total Revenue for FY2024 was $236,584M.
Would you also like to know about Microsoft's net income or revenue growth for FY2024?

Example 2: Growth Metric for a Company
- **User query:** What is Apple's revenue growth in 2023?
- **Chatbot response:** Apple's Total Revenue for FY2023 was $383,285M. Apple's Revenue Growth for FY2023 was -2.80%.
Would you like to know about Apple's other growth metrics or a summary of its financial health for FY2023?

Example 3: General Performance Summary
- **User query:** Summarize Tesla's performance.
- **Chatbot response:** 
Here's a summary of Tesla's financial performance for FY2024:
\- Total Revenue: $97,690M
\- Net Income: $7,153M
\- Cash Flow from Operations: $14,923M
\- Revenue Growth (YoY): 0.95%
\- Net Income Growth (YoY): -52.23%

Is there a specific metric you'd like to dive deeper into, or perhaps compare another year's performance?

Example 4: Querying for an Unsupported Year (Demonstrates Fallback)
- **User query:** What is Google's profit?
- **Chatbot response:** I need a company name (Microsoft, Tesla, or Apple) to provide financial insights. Please try again.

### 5. Limitations
It's important to note the current limitations of this rule-based chatbot:

* **Strict Keyword Matching:** It relies heavily on exact keyword matches and predefined patterns. It cannot understand synonyms, complex sentence structures, or nuanced natural language inputs beyond what is explicitly coded.
* **Limited Conversational Context:** The chatbot does not maintain deep conversational memory. Each query is largely treated independently, meaning it won't remember previous questions or users' implied intent.
* **Fixed Data Scope:** Insights are strictly limited to the financial data provided for Microsoft, Tesla, and Apple for Fiscal Years 2022, 2023, and 2024. It cannot access real-time data, external news, or other companies.
* **Hardcoded Companies:** The list of supported companies (Microsoft, Tesla, Apple) is directly embedded within the function's logic, making it less flexible to add new companies without modifying the function's internal structure.
* **No Advanced AI/ML:** It does not incorporate sophisticated NLP techniques (like intent recognition or entity extraction) or machine learning algorithms for self-improvement or handling unforeseen queries. These advanced features are typically handled by other specialised teams in a larger project context.
* **Text-Based Only:** It does not provide visual aids or integrate with external visualisation tools.

### 6. Conclusion and Variation:
This project demonstrates leveraging structured data with rule-based NLP for a functional financial inquiry chatbot. While limited to explicit rules, it delivers relevant insights, including year-specific data and growth metrics. Future enhancements could include advanced NLP, broader datasets, or visual outputs.

### 7. Key Takeaway:
Effective rule-based chatbots rely on meticulous data preparation and strategic keyword/pattern recognition to deliver accurate, contextually relevant financial information.